### PR TITLE
CAT: Add scenario that OAuth is default method and fields are marked as airbyte_secret

### LINKED ISSUE #33564: Add CAT scenario testing OAuth is the default authentication and oauth fields marked as secret
## Description

As part of certification criteria we should be ensuring that OAuth on cloud is the first option presented to users and that sensitive values are designated as secrets.

## Implementation Details

We can verify this in code by checking the fields in the response returned by the connectorSpecification. We already have a test `test_oauth_flow_parameters()` in `test_core.py`. I think we should be incorporating this into that existing test. Although this is a higher operational certification level, this is a relatively trivial fix even if it were to block community level connectors.

Most connectors follow the pattern of a `credentials` property at the top levelwhich will ultimately contain the OAuth and other auth methods. This can have a `oneOf` is an ordered list and oauth should be the first one. However, we only need to verify ordering if OAuth is specified which is not always the case since not all API support it.

There are some edge cases like:

- source-facebook-marketing/source-facebook-pages: Where `access_token` is at the top level of properties. We don't need to verify because there is no orderering, but this test should not fail either

## Acceptance Criteria

- [x] CAT will fail if a connector's `credentials` has oauth not listed first
- [x] CAT succeeds for connector's that do not specify `credentials`

### PR DESCRIPTION
**Pull request by darynaishchenko**
## What
resolved: https://github.com/airbytehq/airbyte/issues/33564
As part of certification criteria we should be ensuring that OAuth on cloud is the first option presented to users and that sensitive values are designated as secrets.

## How
OAuth is default: Extended existing TestSpec.test_oauth_flow_parameters() with validation for case when credentials object do have an oneOf list and the first method listed is OAuth method.
OAuth fields marked as secret: TestSpec.test_secret_is_properly_marked has been already implemented. Updated unit test for it with OAuth method in test cases spec.

This has been tested on a handful of different sources:
- `source-facebook-pages`
- `source-google-ads`
- `source-facebook-marketing`
- `source-github`

### CODE REVIEW

**Review by tolik0 — APPROVED** [review]

File: airbyte-integrations/bases/connector-acceptance-test/connector_acceptance_test/tests/test_core.py [hunk 1]
```diff
@@ -520,6 +520,24 @@ def test_oauth_flow_parameters(self, actual_connector_spec: ConnectorSpecificati
...
+            credentials = advanced_auth.predicate_key[0]
+            one_of_default_method = None
+            try:
+                one_of_default_method = dpath.util.get(spec_schema, f"/**/{credentials}/oneOf/0")
+            except KeyError:  # Key Error when oneOf is not in credentials object
+                pass
+            if one_of_default_method:
+                path_in_credentials = "/".join(advanced_auth.predicate_key[1:])
+                auth_method_predicate_const = dpath.util.get(one_of_default_method, f"/**/{path_in_credentials}/const")
+                assert (
```
Comment (maxi297) [inline airbyte-integrations/bases/connector-acceptance-test/connector_acceptance_test/tests/test_core.py:538]: I think this assertion will fail for source-zendesk-talk (and maybe others) as [the first element](https://github.com/airbytehq/airbyte/blob/670ffdd32c5869fed3fd8c6d6d9787de84104965/airbyte-integrations/connectors/source-zendesk-talk/source_zendesk_talk/spec.json#L21-L43) is not oauth. How will we get visibility on sources that will fail because of this?

File: airbyte-integrations/bases/connector-acceptance-test/connector_acceptance_test/tests/test_core.py [hunk 2]
```diff
@@ -520,6 +520,24 @@ def test_oauth_flow_parameters(self, actual_connector_spec: ConnectorSpecificati
...
+        If credentials do have oneOf: we check that the OAuth is listed at first.
+        If there is no oneOf and Oauth: OAuth is only option to authenticate the source and no check is needed.
+        """
+        if advanced_auth.predicate_key:
+            credentials = advanced_auth.predicate_key[0]
+            one_of_default_method = None
+            try:
+                one_of_default_method = dpath.util.get(spec_schema, f"/**/{credentials}/oneOf/0")
+            except KeyError:  # Key Error when oneOf is not in credentials object
+                pass
```
Comment (maxi297) [inline airbyte-integrations/bases/connector-acceptance-test/connector_acceptance_test/tests/test_core.py:534]: It would have been nice to keep a trace of this using `pytest.skip`. However, since this test is testing many things, it wouldn't be true that we are skipping the whole test. Is there a reason not to create a new test for that?

File: airbyte-integrations/bases/connector-acceptance-test/unit_tests/test_spec.py [hunk 3]
```diff
@@ -682,6 +682,230 @@ def test_enum_usage(connector_spec, should_fail):
             ),
             "",
         ),
+        # SUCCESS: Fields specified in advanced_auth exist in spec, Oauth is default
```
Comment (maxi297) [inline airbyte-integrations/bases/connector-acceptance-test/unit_tests/test_spec.py:685]: Should we have a test where `predicate_key.oneOf` does not exist in which case, the test should pass?

**Review by maxi297 — COMMENTED** [review]
Just a couple of questions

File: airbyte-integrations/bases/connector-acceptance-test/unit_tests/test_spec.py (hunk 3 above)
Comment (darynaishchenko) [reply airbyte-integrations/bases/connector-acceptance-test/unit_tests/test_spec.py:685]: `predicate_key.oneOf`does not exist  - you mean "credentials" don't have "oneOf" field? This test case is already added https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/bases/connector-acceptance-test/unit_tests/test_spec.py#L628 . The test passes due to  only one option to auth so we don't need to check default.

File: airbyte-integrations/bases/connector-acceptance-test/connector_acceptance_test/tests/test_core.py (hunk 1 above)
Comment (darynaishchenko) [reply airbyte-integrations/bases/connector-acceptance-test/connector_acceptance_test/tests/test_core.py:538]: thanks for pointing this out. i will add possibility to pass skip and bypass reason for this test  in case there is a reason that OAuth is not default method. Health engineer will be responsible to fix sources that fail due to this check.

**Comment by darynaishchenko** [comment]
@maxi297 updated testing flow
- connector health engineer will be responsible for fixing failed connectors
- added skip with bypass reason options that can be added in CAT config

File: airbyte-integrations/bases/connector-acceptance-test/connector_acceptance_test/tests/test_core.py [hunk 4]
```diff
@@ -120,6 +120,12 @@ async def skip_backward_compatibility_tests_fixture(
             pytest.skip(f"Backward compatibility tests are disabled for version {previous_connector_version}.")
         return False
 
+    @pytest.fixture(name="skip_oauth_default_method_test")
+    def skip_oauth_default_method_test_fixture(self, inputs: SpecTestConfig):
+        if inputs.auth_default_method and not inputs.auth_default_method.oauth:
```
Comment (maxi297) [inline airbyte-integrations/bases/connector-acceptance-test/connector_acceptance_test/tests/test_core.py:125]: So if `inputs.auth_default_method` is not defined (which seems possible as [it is optional](https://github.com/airbytehq/airbyte/pull/34178/files#diff-6d30b0c7923582524ce6f15dccff0e22689713a162893c6dcd268c6af46f2302R64)), we will assume oauth is the default method, is that right? Is this a fair assumption? I'm afraid that it will force all the connectors to define `inputs.auth_default_method.oauth == False` which would mean that this is a breaking change. I'm not sure how we usually manage breaking changes for CATs. Do you have more context on that?

**Review by maxi297 — COMMENTED** [review]
Just one clarification and I'm good to go

File: airbyte-integrations/bases/connector-acceptance-test/connector_acceptance_test/tests/test_core.py (hunk 4 above)
Comment (darynaishchenko) [reply airbyte-integrations/bases/connector-acceptance-test/connector_acceptance_test/tests/test_core.py:125]: Yes, when `auth_default_method` is not defined in `acceptance-test-config.yml` we assume that OAuth is default method. I'm not sure what you mean about a breaking change. If there is a weigh reason for some connector to do not specify Oauth as default engineer can add `auth_default_method.oauth == False` to cat config. Otherwise, engineer should change the order in oneOf in source connector spec  which is not a breaking change for source.

Comment (maxi297) [reply airbyte-integrations/bases/connector-acceptance-test/connector_acceptance_test/tests/test_core.py:125]: > If there is a weigh reason for some connector to do not specify Oauth as default engineer can add auth_default_method.oauth == False

So will there be a follow up work to fix all the CATs that will fail because of this? For example, source-stripe does not suppor oauth. This change will break CATs for this connector. What is the plan to upgrade all connectors like source-stripe? Or should we make this change non-breaking?

Comment (darynaishchenko) [reply airbyte-integrations/bases/connector-acceptance-test/connector_acceptance_test/tests/test_core.py:125]: according to this comment https://github.com/airbytehq/airbyte/issues/33564#issuecomment-1863272005 I updated existing CAT scenatio, but there is a way to make this change non-breaking: update TestConnectorAttributes class and this test will not work unless config for it will be added to acceptance-test-config.yml.

**Review by lazebnyi — APPROVED** [review]

File: airbyte-integrations/bases/connector-acceptance-test/connector_acceptance_test/tests/test_core.py (hunk 4 above)
Comment (brianjlai) [reply airbyte-integrations/bases/connector-acceptance-test/connector_acceptance_test/tests/test_core.py:125]: I'll take some time to review this after some morning meetings, but first thoughts on this laying out the possible scenarios:

1. Connector only supports API key -> The test is skipped regardless of CAT config
2. Connector supports OAuth and API key, the OAuth is default -> Passes test
3. Connector supports OAuth and API key, API key is default -> Fails test
4. Connector supports OAuth and API key, API key is default, override is enabled-> Skips/passes test.

> This is a breaking change and if your connector does not support oauth, you need to define inputs.auth_default_method.oauth == False

I think we want to try to avoid this because its pretty involved to change many of our connectors.

Basically, I think we want to get to a place where connectors that don't even support OAuth should not need to define anything in the CAT spec and we skip the test entirely because it's irrelevant. I need to read through this in more detail, but given we have the whole spec, I would think we can figure out if OAuth is one of the auth methods.

And for any source that does have an OAuth as an option, then we need to verify it comes first, unless the override is specified. I think even those this might break existing Silver certified connectors, its a valid test failure because it means we forgot to fix this during the certification

**Review by brianjlai — COMMENTED** [review]
two thoughts here:

1.
Do we need a separate CAT acceptance config at all? Rather than force each connector to explicitly define this field, could we just always run the test no matter what connector? And we either skip or pass the test if a connector's spec doesn't have any form of OAuth to validate.

And for the connectors that do have OAuth in the spec, then we verify it's the first option. That seems like the least impact on our existing connectors. And if we're concerned about this test failing for uncertified connectors, we can use `operational_certification_test` fixture to skip the test when it quality level is not at the right value

2.
I'm curious if we have seen any use cases for why OAuth shouldn't be first option and it requires a bypass reason?

**Comment by darynaishchenko** [comment]
@brianjlai , For current implementation a separate CAT acceptance config is optional, the test  always runs as part of TestSpec scenario. The test is skipped in case 1. no oauth in connector. 2. no oneOf (only one way to auth), if oauth is present  3. auth_default_method.oauth == False with bypass reason provided in  CAT acceptance config (A connector health engineer(secondary oncall) will be responsible to fix oauth default method or add skip and bypass in cat config.) 

> Do we need a separate CAT acceptance config at all? Rather than force each connector to explicitly define this field, could we just always run the test no matter what connector? And we either skip or pass the test if a connector's spec doesn't have any form of OAuth to validate.

It's no problem to add operational_certification_test fixture, but the reason why it wasn't added was mentioned in the task https://github.com/airbytehq/airbyte/issues/33564 (Although this is a higher operational certification level, this is a relatively trivial fix even if it were to block community level connectors.) For example zendesk-talk is certified and generally_available but OAuth is not default option, I didn't find  the reason in certification checklist, it needs more investigation that a secondary oncall can do while his work on connector health and provide a fix/bypass.

> And for the connectors that do have OAuth in the spec, then we verify it's the first option. That seems like the least impact on our existing connectors. And if we're concerned about this test failing for uncertified connectors, we can use `operational_certification_test` fixture to skip the test when it quality level is not at the right value

File: airbyte-integrations/bases/connector-acceptance-test/connector_acceptance_test/config.py [hunk 5]
```diff
@@ -42,6 +42,17 @@ class BackwardCompatibilityTestsConfig(BaseConfig):
     )
 
 
+class OAuthTestConfig(BaseConfig):
+    oauth = Field(True, description="Allow source to have another default method that OAuth.")
+    bypass_reason: Optional[str] = Field(description="Reason why OAuth is not default method.")
+
+    @validator("oauth", always=True)
+    def validate_exact_order(cls, oauth, values):
```
Comment (brianjlai) [inline airbyte-integrations/bases/connector-acceptance-test/connector_acceptance_test/config.py:50]: why is this called `validate_exact_order`? This should be `oauth` or something presumably

**Review by brianjlai — APPROVED** [review]
One naming correction on the config validator and there's the merge conflict on the CHANGELOG. Everything else looks good to me.

 I think we're in alignment that it will just incorporated w/o the QL taken into account and primary/secondary can fix them if the do come up

File: airbyte-integrations/bases/connector-acceptance-test/connector_acceptance_test/config.py (hunk 5 above)
Comment (darynaishchenko) [reply airbyte-integrations/bases/connector-acceptance-test/connector_acceptance_test/config.py]: updated

**Comment by darynaishchenko** [comment]
@maxi297, I have approve from @brianjlai. Is everything good from your side?
